src/common/Journald.cc

```diff
@@ -0,0 +1,280 @@
+    thread_key.copy(m_thread, thread_key.size());
+    message_key.copy(m_message_key, message_key.size());
+  }
+
+  constexpr struct iovec *iovec() { return this->m_msg_vec; }
+  constexpr std::size_t iovec_len() { 
+    return sizeof(m_msg_vec) / sizeof(m_msg_vec[0]);
+  }
+
+  void encode(const Entry& e, const SubsystemMap *s) {
+    int ret;
+
+    // m_msg_vec[0]
+    m_priority_subsys[priority_key.size()] = '0' + map_prio(e.m_prio);
+
+    auto subsys_name = s->get_name(e.m_subsys);
+    m_msg_vec[1].iov_base = (void *)subsys_name;
+    m_msg_vec[1].iov_len = strlen(subsys_name);
+
+    ret = snprintf(m_timestamp + timestamp_key.size(), max_timestamp_len,
+      "%lu", e.m_stamp.time_since_epoch().count().count);
+    ceph_assert(static_cast<std::size_t>(ret) < max_timestamp_len);
+    m_msg_vec[2].iov_len = timestamp_key.size() + ret;
+
+    ret = snprintf(m_prio + prio_key.size(), max_prio_len, "%hd", e.m_prio);
+    ceph_assert(static_cast<std::size_t>(ret) < max_prio_len);
+    m_msg_vec[3].iov_len = prio_key.size() + ret;
+
+    // m_msg_vec[4]
+    ret = snprintf(m_thread + thread_key.size(), max_thread_len, "%016lx", e.m_thread);
+    ceph_assert(ret == max_thread_len - 1);
+
+    *reinterpret_cast<uint64_t*>(m_message_key+message_key.size()) = htole64(e.size());
+    m_msg_vec[6].iov_base = (void *)e.strv().data();
+    m_msg_vec[6].iov_len = e.size();
+  }
+ private:
+  static constexpr std::string_view priority_key  = "PRIORITY=";
+  static constexpr std::string_view subsys_key    = "\nCEPH_SUBSYS=";
+  static constexpr std::string_view timestamp_key = "\nTIMESTAMP=";
+  static constexpr std::string_view prio_key      = "\nCEPH_PRIO=";
+  static constexpr std::string_view thread_key    = "\nTHREAD=";
+  static constexpr std::string_view message_key   = "\nMESSAGE\n";
+
+  static constexpr std::size_t max_timestamp_len =
+    std::numeric_limits<uint64_t>::digits10 + 2; /* round up, '\0' */
+  static constexpr std::size_t max_prio_len =
+    std::numeric_limits<short>::digits10 + 3; /* round up, minus sign, '\0' */
+  static constexpr std::size_t max_thread_len =
+    std::numeric_limits<pthread_t>::digits / 4 + 1; /* '\0' */
+
+  static constexpr auto subsys_key_offset = priority_key.size() + 1;
+
+  struct iovec m_msg_vec[8];
+  char m_priority_subsys[
+    priority_key.size() + 1 +
+    subsys_key.size()];
+  char m_timestamp[timestamp_key.size() + max_timestamp_len];
+  char m_prio[prio_key.size() + max_prio_len];
+  char m_thread[thread_key.size() + max_thread_len];
+  char m_message_key[message_key.size() + sizeof(uint64_t)];
+};
+
+class LogEntryEncoder {
+ public:
+  LogEntryEncoder() :
+    m_msg_vec {
+      {}, {}, {.iov_base = (char *)"\n", .iov_len=1},
+    } {}
+
+  constexpr struct iovec *iovec() { return this->m_msg_vec; }
+  constexpr std::size_t iovec_len() { 
+    return sizeof(m_msg_vec) / sizeof(m_msg_vec[0]);
+  }
+
+  void encode(const LogEntry& le) {
+    std::stringstream meta;
```

Comment:
i'd avoid using `stringstream` in this case as the performance hurts if we construct it every time we encode a `LogEntry`. instead, i'd suggest use libfmt's in memory writer. like
```c++
#include <fmt/format.h>
#include <fmt/ostream.h>
///...
      fmt::format_to(meta_buf,
             R"(PRIORITY={}
TIMESTAMP={}
CEPH_NAME={}
CEPH_RANK={}
CEPH_SEQ={}
CEPH_CHANNEL={}
MESSAGE
)",
             clog_type_to_syslog_level(le.prio),
             le.stamp.to_nsec(),
             le.name.to_str(),
             le.rank,
             le.seq,
             le.channel);
      uint64_t msg_len = htole64(le.msg.size());
      meta_buf.append(&msg_len, &msg_len + 1);
//...
private:
  fmt::memory_buffer meta_buf;
```